src/transform/src/predicate_pushdown.rs

```diff
@@ -385,6 +385,29 @@ impl PredicatePushdown {
                         }
                         *relation = result;
                     }
+                    MirRelationExpr::FlatMap {
+                        input, func, exprs, ..
+                    } => {
+                        // println!("{:}", "enter MirRelationExpr::FlatMap case");
+                        // println!("{:}", predicates.iter().map(|x| x.to_string()).join(","));
+                        let (retained, pushdown) = Self::push_filters_through_flat_map(
+                            predicates,
+                            input.arity(),
+                            all_errors,
+                        );
+                        let mut result = input.take_dangerous();
```

Comment:
Here you could just do:

```
std::mem::swap(&mut retained, predicates);

if !pushdown.is_empty()  {
    // put the filter on top of the input
    **input = input.take_dangerous().filter(pushdown);
}

// ... and keep pushing predicates down
self.action(input, get_predicates);
```

`fusion::Filter` transform will be then responsible for removing the empty `Filter` operator.

Reply:
Done.